python/python_reduction.cc

```diff
@@ -0,0 +1,358 @@
+  {
+    try
+    {
+      return this->py_reduction_impl->attr(method_name)(std::forward<Args>(args)...);
+    }
+    catch (...)
+    {
+      // TODO: Properly translate and return Python exception. #2169
+      PyErr_Print();
+      PyErr_Clear();
+      THROW("Exception when calling into python method: " << method_name);
+    }
+  }
+
+  // todo review if dont_delete_me is still needed with this refactor
+  void actual_learn(example* ec)
+  {
+    this->call_py_impl_method("_learn_convenience", example_ptr(ec, dont_delete_me),
+        py_cpp_callback_ptr(new py_cpp_callback(base_learner), dont_delete_me));
+  }
+
+  void actual_predict(example* ec)
+  {
+    this->call_py_impl_method("_predict_convenience", example_ptr(ec, dont_delete_me),
+        py_cpp_callback_ptr(new py_cpp_callback(base_learner), dont_delete_me));
+  }
+
+  void actual_finish_example(example* ec)
+  {
+    this->call_py_impl_method("_finish_example", example_ptr(ec, dont_delete_me));
+  }
+
+  ex_list multi_ex_to_boost(multi_ex* examples)
+  {
+    ex_list list;
+    for (auto ec : *examples) { list.emplace_back(ec, dont_delete_me); }
+
+    return list;
+  }
+
+  void actual_learn(multi_ex* examples)
+  {
+    ex_list list = multi_ex_to_boost(examples);
+    this->call_py_impl_method(
+        "_learn_convenience", list, py_cpp_callback_ptr(new py_cpp_callback(base_learner, examples), dont_delete_me));
+  }
+
+  void actual_predict(multi_ex* examples)
+  {
+    ex_list list = multi_ex_to_boost(examples);
+    this->call_py_impl_method(
+        "_predict_convenience", list, py_cpp_callback_ptr(new py_cpp_callback(base_learner, examples), dont_delete_me));
+  }
+
+  void actual_finish_example(multi_ex* examples)
+  {
+    ex_list list = multi_ex_to_boost(examples);
+    this->call_py_impl_method("_finish_example", list);
+  }
+
+  void actual_saveload(io_buf* model_file, bool read, bool text)
+  {
+    this->call_py_impl_method(
+        "_save_load_convenience", read, text, py_cpp_callback_ptr(new py_cpp_callback(model_file)));
+  }
+
+  void set_base_learner(void* learner) { this->base_learner = reinterpret_cast<VW::LEARNER::base_learner*>(learner); }
+};
+
+// specialization needed to compile, this should never be reached since we always use
+// VW::config::supported_options_types
+template <>
+py::object option_manager::base_option_to_pyobject<VW::config::typelist<>>(VW::config::base_option& options)
+{
+  return py::object();
+}
+
+struct custom_builder_with_binding : VW::default_reduction_stack_setup
+{
+  std::unique_ptr<wrapper::external_binding> instance;
+
+  bool should_call_custom_python_setup = false;
+  std::string call_after_executing;
+
+  // decide here if its base, multi or single
+  custom_builder_with_binding(std::unique_ptr<wrapper::external_binding> _instance)
+  {
+    instance = std::move(_instance);
+    call_after_executing = "extra_metrics";
+
+    // if (is_singleline) call_after_executing = "extra_metrics";
```

Comment:
we don't want these checks?

Reply:
we do, but I'll try to merge only with a normal reduction and add support for customizing the type of reduction later. i.e. ask it to be base, ask it to be a multi, etc.